frame/lending/src/tests_offchain.rs

```diff
@@ -0,0 +1,159 @@
+		// from 50_000 USDT to 38_000 USDT.
+		// Now the borrow is undercolateraized since market's collateral factor equals two.
+		// Therefore, one BTC can cover only 19_000 of 20_0000 borrowed USDT.
+		set_price(BTC::ID, NORMALIZED::units(38_000));
+		// Header for the fake block to execute off-chain worker
+		let header =
+			Header::new(2, H256::default(), H256::default(), [69u8; 32].into(), Digest::default());
+		// Execute off-chain worker
+		Executive::offchain_worker(&header);
+		// Check if corresponded liquidation transaction has been placed in the pool
+		let tx = pool_state.write().transactions.pop().unwrap();
+		assert!(pool_state.read().transactions.is_empty());
+		let tx = Extrinsic::decode(&mut &*tx).unwrap();
+		assert_eq!(
+			tx.call,
+			Call::Lending(crate::Call::liquidate { market_id, borrowers: vec![manager.clone()] })
+		);
+		process_block_with_execution(tx);
+		// Check event from Lending pallet
+		let event =
+			crate::Event::LiquidationInitiated { market_id, borrowers: vec![manager.clone()] };
+		System::assert_has_event(Event::Lending(event));
+		// Check event from Liquidations pallet
+		let event = pallet_liquidations::Event::PositionWasSentToLiquidation {};
+		System::assert_has_event(Event::Liquidations(event));
+	});
+}
+
+// HELPERS
+fn default_under_collateralized_warn_percent() -> Percent {
+	Percent::from_float(0.10)
+}
+
+fn create_market<const NORMALIZED_PRICE: u128>(
+	borrow_asset: RuntimeCurrency,
+	collateral_asset: RuntimeCurrency,
+	manager: AccountId,
+	reserved_factor: Perquintill,
+	collateral_factor: MoreThanOneFixedU128,
+) -> (MarketIndex, VaultId) {
+	set_price(borrow_asset.id(), NORMALIZED::ONE);
+	set_price(collateral_asset.id(), NORMALIZED::units(NORMALIZED_PRICE));
+
+	Tokens::mint_into(borrow_asset.id(), &manager, borrow_asset.units(1000)).unwrap();
+	Tokens::mint_into(collateral_asset.id(), &manager, collateral_asset.units(100)).unwrap();
+
+	let config = CreateInput {
+		updatable: UpdateInput {
+			collateral_factor,
+			under_collateralized_warn_percent: default_under_collateralized_warn_percent(),
+			liquidators: vec![],
+			interest_rate_model: InterestRateModel::default(),
+		},
+		reserved_factor,
+		currency_pair: CurrencyPair::new(collateral_asset.id(), borrow_asset.id()),
+	};
+
+	Lending::create_market(Origin::signed(manager), config.try_into_validated().unwrap()).unwrap();
+
+	let system_events = System::events();
+	if let Some(EventRecord {
+		event:
+			Event::Lending(crate::Event::<Runtime>::MarketCreated {
+				market_id,
+				vault_id,
+				manager: _,
+				currency_pair: _,
+			}),
+		..
+	}) = system_events.last()
+	{
+		(*market_id, *vault_id)
+	} else {
+		panic!(
+			"System::events() did not contain the market creation event. Found {:#?}",
+			system_events
+		)
+	}
+}
+
+fn assert_extrinsic_event<T: crate::Config>(
+	result: DispatchResultWithPostInfo,
+	event: <T as crate::Config>::Event,
+) {
+	assert_ok!(result);
+	assert_last_event::<T>(event);
+}
```

Comment:
these can be imported from `crate::tests` instead of being redefined

Reply:
Done, fn create_market was generalized. 
Please be aware that now it does not emit events.

Reply:
It's possible to make it emit events - just be generic over the runtime instead of the pallet type. It will require a bit of refactoring, I can message you on slack if you want to peer program it

Reply:
the basic idea is this: ```rust
crate::Pallet::<Runtime>::create_market(manager, validated.value())
```

Reply:
@benluelo Done